### PR TITLE
Output: Preload the first page’s background media

### LINKED ISSUE #6742: Preload important resources
## Feature Description

Add [preloading](https://web.dev/optimize-lcp/?utm_source=lighthouse&utm_medium=lr#preload-important-resources) functionality to the plugin to improve loading time. 

As an example, this [story URL](https://www.selectgame.com.br/web-stories/esquadrao-omega-em-league-of-legends-wild-rift/) performance [report](https://lighthouse-dot-webdotdevsite.appspot.com//lh/html?url=https%3A%2F%2Fwww.selectgame.com.br%2Fweb-stories%2Fesquadrao-omega-em-league-of-legends-wild-rift%2F#performance) lists an opportunity to preload the image to improve the LCP time, which the plugin can add the preload attribute to improve this.

## Additional Context

Related support topic: https://wordpress.org/support/topic/about-page-speed-insights-tests/

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

LCP for stories could be improved through the right uses of preloading.

## Summary

Adds `link[preload]` elements in AMP HTML output.

Right now, this only includes the first page's background image/video/GIF and nothing else.

## Relevant Technical Choices

* Does not add `Link` HTTP headers as the benefit seems unnecessary and none of the other preloaded AMP scripts are added this way.

## To-do

Should more elements be included here, e.g. other large elements on the first page?

## User-facing changes

N/A

## Testing Instructions

1. Set an image/video/GIF as the first page's background
2. Preview story
3. Verify that a `link[preload]` element has been added for that image

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6742

### CODE REVIEW

**Comment by spacedmonkey** [comment]
For some reason I thought this was going to be done in PHP...

**Comment by swissspidy** [comment]
> For some reason I thought this was going to be done in PHP...

Turning the HTML tags into HTTP headers could be done in PHP, but I don‘t see a good enough reason for doing that

**Comment by gmajoulet** [comment]
Just curious, have you been running some performance testing to see if this actually makes things better? We're planning on looking into this and probably adding an automatic transformation for the AMP cache version of each Story, but we'll be testing whether it's making things worse or better first.

Doing some research on the topic I found [this tweet](https://twitter.com/shubhie/status/1391891309338660864) that links to [this doc](https://docs.google.com/document/d/1ZEi-XXhpajrnq8oqs5SiW-CXR3jMc20jWIzN5QRy1QA/edit) explaining how `preload` can hurt performance, originating from Google.

**Comment by swissspidy** [comment]
> Just curious, have you been running some performance testing to see if this actually makes things better? We're planning on looking into this and probably adding an automatic transformation for the AMP cache version of each Story, but we'll be testing whether it's making things worse or better first.
> 
> Doing some research on the topic I found [this tweet](https://twitter.com/shubhie/status/1391891309338660864) that links to [this doc](https://docs.google.com/document/d/1ZEi-XXhpajrnq8oqs5SiW-CXR3jMc20jWIzN5QRy1QA/edit) explaining how `preload` can hurt performance, originating from Google.

Not yet, but definitely warrants some testing first. Definitely a good callout!

**Comment by vladoa** [comment]
Hey, just curious, what is the plan for this? Will it be merged? Meanwhile all web stories are showing LCP issue in Lighthouse.

**Comment by swissspidy** [comment]
> Hey, just curious, what is the plan for this? Will it be merged? Meanwhile all web stories are showing LCP issue in Lighthouse.

Thanks for your interest!

As per the above comments, preloading can be counterproductive when not done right, which is why we‘re going to do some testing for this first. If you want to give this development build a try on a staging site and give us some feedback, that would be great.

That being said, you might also want to check out our latest 1.7.1 release, where we leverage AMP‘s ESM builds for improved performance on modern browsers, positively affecting web vitals scores.

File: assets/src/edit-story/output/utils/getPreloadResources.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * @return {Array<PreloadResource>} List of preload resources.
+ */
+function getPreloadResources(pages) {
+  const preloadResources = [];
+
+  if (pages.length === 0) {
+    return preloadResources;
+  }
+
+  for (const { type, resource, isBackground } of pages[0].elements) {
```
Comment (spacedmonkey) [inline assets/src/edit-story/output/utils/getPreloadResources.js:40]: Should this be just the first page or the first page that has rich media?

Comment (swissspidy) [reply assets/src/edit-story/output/utils/getPreloadResources.js:40]: Great question!

Just the first page overall. Media on any pages beyond that will be preloaded by the AMP runtime as the user navigates through the story.

**Comment by spacedmonkey** [comment]
I did some testing and there some seem to be very small benefit.  2% is something. 
### Before
Screenshot 2021-06-07 at 14 55 14

### After 
Screenshot 2021-06-07 at 14 36 26

As other web stories editor as using it. See https://github.com/google/web-stories-wp/issues/6742#issuecomment-855928006 . I think this is a safe change.

**Comment by swissspidy** [comment]
Dropping LCP from 1.1s to 0.7s with this change sounds like a nice win indeed! 👍

**Comment by gmajoulet** [comment]
> Dropping LCP from 1.1s to 0.7s

!!!!
We should take this with a grain of salt as this is local testing, development build etc, but this is extremely promising. We'll do on our testing very very soon as well
cc @mszylkowski

**Comment by swissspidy** [comment]
> > Dropping LCP from 1.1s to 0.7s
> 
> !!!!
> We should take this with a grain of salt as this is local testing, development build etc, but this is extremely promising. We'll do on our testing very very soon as well


For sure 👍 

@choumx WDYT about adding this to 1.8.0?

This of course won't affect existing stories until they're saved again.

**Review by spacedmonkey — APPROVED** [review]
LGTM

**Comment by gmajoulet** [comment]
@swissspidy @choumx let's get in touch to see how we can track the LCP improvements from this feature?

**Pull request merged**